pygame_gui/elements/ui_auto_resizing_container.py

```diff
@@ -0,0 +1,258 @@
+        self.right_to_left_elements = sorted(elements, key=(lambda element: element.get_abs_rect().right), reverse=True)
+        self.top_to_bottom_elements = sorted(elements, key=(lambda element: element.get_abs_rect().top))
+        self.bottom_to_top_elements = sorted(elements, key=(lambda element: element.get_abs_rect().bottom),
+                                             reverse=True)
+
+    def _update_extreme_elements(self) -> None:
+        """
+        Updates which elements is currently responsible for preventing the container to collapse from a certain side
+        :return: None
+        """
+        self.left_element = next(
+            (element for element in self.left_to_right_elements if element not in self.left_anchor_elements), None)
+        self.right_element = next(
+            (element for element in self.right_to_left_elements if element not in self.right_anchor_elements), None)
+        self.top_element = next(
+            (element for element in self.top_to_bottom_elements if element not in self.top_anchor_elements), None)
+        self.bottom_element = next(
+            (element for element in self.bottom_to_top_elements if element not in self.bottom_anchor_elements), None)
+
+    def _update_dimensions(self) -> None:
+        """
+        Updates the the sizes of the container
+        :return: None
+        """
+        left_ext = self.get_abs_rect().left - self.get_left_most_point() if self.resize_left else 0
+        right_ext = self.get_right_most_point() - self.get_abs_rect().right if self.resize_right else 0
+        top_ext = self.get_abs_rect().top - self.get_top_most_point() if self.resize_top else 0
+        bottom_ext = self.get_bottom_most_point() - self.get_abs_rect().bottom if self.resize_bottom else 0
+
+        if left_ext or top_ext:
+            self.set_position(pygame.Vector2(self.get_abs_rect().topleft) - pygame.Vector2(left_ext, top_ext))
+
+        width = left_ext + self.get_abs_rect().width + right_ext
+        height = top_ext + self.get_abs_rect().height + bottom_ext
+
+        if left_ext or right_ext or top_ext or bottom_ext:
+            self.set_dimensions((width, height))
+
+    def update(self, time_delta: float) -> None:
+        """
+        Updates the container's size based upon the elements inside.
+
+        Call this function if you have added or removed an element from this container and want
+        to update the size in the same frame, otherwise it will update in the next frame.
+
+        :param time_delta: The time passed between frames, measured in seconds.
+        :return: None
+        """
+        super().update(time_delta)
+        # Perhaps this could be attached to a timer to not have to call every single frame
+        self._update_anchor_elements()
+        self._update_sorting()
+        self._update_extreme_elements()
+        self._update_dimensions()
```

Comment:
Instead of doing this - checking the container's contents every loop (which could have large performance implications in a container with lots of elements inside), I think we should update these things only when we know the container may need to change size.

As far as I can tell this is in one of four situations:

1. A new element has been added to the container.
2. An existing element has been removed form the container.
3. An element already inside the container has changed position.
4. An element within the container has changed size.

There is already a function on all containers that all elements call when they change size or move, it is called `on_anchor_target_changed()`. I would rename this function in the `UIContainer()` to something more generic like `on_content_moved_or_resized()` and then override it in this class and have it do this work.

For adding and removing you can just override the existing `add_element())` and `remove_element()`. Once you've covered all four you should be able to remove this update function I would think?